### DISCUSSION TITLE
Using document extensions with REST API

### DISCUSSION THREAD

**Discussion by roqvist** [ravendb/ravendb#14406] [Q&A] [answered]
I'm trying to find information on how to use document extensions (counters and time series specifically) with the REST API. These features seem to be documented in various client APIs only. Does anyone know?

**Comment by ayende** [comment, marked as answer]
Here are some examples:

`GET http://live-test.ravendb.net/databases/DemoUser-9d8b46cd-4956-47d0-ab30-6f12d8b6a368/counters?docId=products/8-A`

Timeseries:

`GET http://live-test.ravendb.net/databases/DemoUser-9d8b46cd-4956-47d0-ab30-6f12d8b6a368/timeseries?docId=companies/9-A&name=StockPrices`

What is your scenario for doing this via REST directly, however?

**Reply by roqvist** [reply]
I'm using Rust

**Reply by ayende** [reply]
In that case, I would suggest that you'll copy the `Operations & Commands` model from the API.
That way, you could integrate new commands far more easily, and can usually just copy the structures and have them work.

We would like a Rust client, but as you can imagine, a full blown client is a _big_ thing, given the scope of RavenDB.
If you are interested, we'll be very happy to assist with that.

**Reply by roqvist** [reply]
Thanks for the suggestion, I will take a look. For the prototyping and evaluation I'm currently working on simply incrementing counters will suffice, but there will obviously be some abstractions over the REST calls and also some simple caching.

API clients can be a lot of work, I develop and maintain [one myself as a hobby project](https://crates.io/crates/imagevault) and even a simple REST API has a lot to take into consideration. The good thing is Rust has a very nice eco system around networking so a lot of the heavy lifting is already available for use. It basically boils down to designing a simple and expressive enough structure for the API.

If/when considering a Raven Rust client, I would love the ability to compile for embedded devices. Raven's counters and time series extensions would be a great fit for micro controllers/IOT projects. (Typically placing any platform specific stuff as [features ](https://doc.rust-lang.org/cargo/reference/features.html), that can be disabled when targeting embedded devices).

**Reply by ayende** [reply]
Building a fully fledged client is a big task, yes. I wonder if `crust` can help here.

That said, there are actually three separate levels of a RavenDB client.

* Provide strongly typed operations & commands. That is mostly about representing the input / output from the server in a nice manner.
* Provide failover behaviors - that is about reading and understanding the ravendb topology and failing appropraitely
* Strong client - unit of work, change tracking, etc.

The first two stages are fairly simple. For example, here is the basic details in PHP: https://gist.github.com/ayende/ba2abe28fbb5fed3004cbdedaadac393

**Reply by ayende** [reply]
About embedded, we have users that make use of the C++ client on ARM32 instances deployed in the field.
